src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -60,15 +86,54 @@ static void WriteAsset(ObjectWriter writer, ISerializerService serializer, Check
             }
         }
 
-        public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(Stream stream, int scopeId, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
+        public static ValueTask<ImmutableArray<(Checksum, object)>> ReadDataAsync(PipeReader pipeReader, int scopeId, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
         {
-            using var _ = ArrayBuilder<(Checksum, object)>.GetInstance(out var results);
+            // Workaround for ObjectReader not supporting async reading.
+            // Unless we read from the RPC stream asynchronously and with cancallation support we might hang when the server cancels.
+            // https://github.com/dotnet/roslyn/issues/47861
```

Comment:
trying to understand this better.  is this because the server canceling doesn't interrupt the connection/pipe/whatever?  trying to understand if this is an issue on our end, or in components we use (and if so, which ones).

Reply:
> is this because the server canceling doesn't interrupt the connection/pipe/whatever? 

yes.

Reply:
> trying to understand if this is an issue on our end, or in components we use (and if so, which ones).

Reply:
It is on our end. We should not use Stream.Read/Write that are blocking and can't be canceled.

Reply:
Ok, thanks for clarifying that :)